### PR TITLE
Feature: support baichuan serial models, by now, including Baichuan-7…

### PR DESCRIPTION
**Pull request by jameswu2014**
As more and more people begin to use Baichuan's open-source models, the influence of Baichuan models is growing, especially in China. Many community members are interested in adding support for Baichuan models to llama.cpp. Meanwhile, Baichuan is a very open company, and in the future, it plans to open-source more and more models, taking all this into consideration, we would like to add support for the Baichuan model to llama.cpp. To do this, we need to make some changes, which we hope can be merged into the main branch of llama.cpp. In the future, we would be happy to help maintain support for Baichuan models in llama.cpp. We sincerely hope that our pull request can be accepted. Thank you.

By the way, the changes of this time mainly for supporting Baichuan-7B and Baichuan-13B, and the future version.

### CODE REVIEW

**Comment by ggerganov** [comment]
Cool!

Lets get some feedback if everything runs smoothly and we can merge

**Comment by jameswu2014** [comment]
> 

OK, Looking forward to the feedback.

**Comment by yinguobing** [comment]
I've tried with this PR and had encountered an model conversion issue with model [Baichuan-13B-Chat](https://huggingface.co/baichuan-inc/Baichuan-13B-Chat):

```bash
# Error message
Can not map tensor 'model.layers.15.self_attn.W_pack.weight'
```

After a little investigation, it seems [this line of code](https://github.com/jameswu2014/llama.cpp/blob/37873ae77ae114d8cc28ebef017520278037ad95/convert-baichuan-hf-to-gguf.py#L243) will break the model modification loop on layer 15, which is an early stop as the left parts have similar layers should be modified:
```python
# Original code
    for i in itertools.count():
        if f"model.layers.{i}.self_attn.W_pack.weight" in model_part:
            print(f"Unpacking and permuting layer {i}")
            tmp[f"model.layers.{i}.self_attn.q_proj.weight"]=reverse_hf_permute_part(model_part[f"model.layers.{i}.self_attn.W_pack.weight"],0,head_count,head_count)
            tmp[f"model.layers.{i}.self_attn.k_proj.weight"]=reverse_hf_permute_part(model_part[f"model.layers.{i}.self_attn.W_pack.weight"],1,head_count,head_count_kv)
            tmp[f"model.layers.{i}.self_attn.v_proj.weight"]=reverse_hf_part(model_part[f"model.layers.{i}.self_attn.W_pack.weight"],2)
            del tmp[f"model.layers.{i}.self_attn.W_pack.weight"]
        else:
            continue # <- Breaks on layer 15
```

A possible fix:
```python
    for i in range(block_count):
        if f"model.layers.{i}.self_attn.W_pack.weight" in model_part:
            print(f"Unpacking and permuting layer {i}")
            tmp[f"model.layers.{i}.self_attn.q_proj.weight"]=reverse_hf_permute_part(model_part[f"model.layers.{i}.self_attn.W_pack.weight"],0,head_count,head_count)
            tmp[f"model.layers.{i}.self_attn.k_proj.weight"]=reverse_hf_permute_part(model_part[f"model.layers.{i}.self_attn.W_pack.weight"],1,head_count,head_count_kv)
            tmp[f"model.layers.{i}.self_attn.v_proj.weight"]=reverse_hf_part(model_part[f"model.layers.{i}.self_attn.W_pack.weight"],2)
            del tmp[f"model.layers.{i}.self_attn.W_pack.weight"]
```

The final result:
baichuan-13b-chat

**Comment by jameswu2014** [comment]
> I've tried with this PR and had encountered an model conversion issue with model [Baichuan-13B-Chat](https://huggingface.co/baichuan-inc/Baichuan-13B-Chat):
> 
> ```shell
> # Error message
> Can not map tensor 'model.layers.15.self_attn.W_pack.weight'
> ```
> 
> After a little investigation, it seems [this line of code](https://github.com/jameswu2014/llama.cpp/blob/37873ae77ae114d8cc28ebef017520278037ad95/convert-baichuan-hf-to-gguf.py#L243) will break the model modification loop on layer 15, which is an early stop as the left parts have similar layers should be modified:
> 
> ```python
> # Original code
>     for i in itertools.count():
>         if f"model.layers.{i}.self_attn.W_pack.weight" in model_part:
>             print(f"Unpacking and permuting layer {i}")
>             tmp[f"model.layers.{i}.self_attn.q_proj.weight"]=reverse_hf_permute_part(model_part[f"model.layers.{i}.self_attn.W_pack.weight"],0,head_count,head_count)
>             tmp[f"model.layers.{i}.self_attn.k_proj.weight"]=reverse_hf_permute_part(model_part[f"model.layers.{i}.self_attn.W_pack.weight"],1,head_count,head_count_kv)
>             tmp[f"model.layers.{i}.self_attn.v_proj.weight"]=reverse_hf_part(model_part[f"model.layers.{i}.self_attn.W_pack.weight"],2)
>             del tmp[f"model.layers.{i}.self_attn.W_pack.weight"]
>         else:
>             continue # <- Breaks on layer 15
> ```
> 
> A possible fix:
> 
> ```python
>     for i in range(block_count):
>         if f"model.layers.{i}.self_attn.W_pack.weight" in model_part:
>             print(f"Unpacking and permuting layer {i}")
>             tmp[f"model.layers.{i}.self_attn.q_proj.weight"]=reverse_hf_permute_part(model_part[f"model.layers.{i}.self_attn.W_pack.weight"],0,head_count,head_count)
>             tmp[f"model.layers.{i}.self_attn.k_proj.weight"]=reverse_hf_permute_part(model_part[f"model.layers.{i}.self_attn.W_pack.weight"],1,head_count,head_count_kv)
>             tmp[f"model.layers.{i}.self_attn.v_proj.weight"]=reverse_hf_part(model_part[f"model.layers.{i}.self_attn.W_pack.weight"],2)
>             del tmp[f"model.layers.{i}.self_attn.W_pack.weight"]
> ```
> 
> The final result: baichuan-13b-chat

Thanks for your feedback, you are right, we fixed it.

**Comment by yinguobing** [comment]
Besides the above fix, it would be better to provide a sample PROMPT file for Baichuan-13B-Chat.

Prompt file: `./prompts/chat-with-baichuan.txt`
```
以下内容为人类用户与与一位智能助手的对话。

用户:你好！
助手:
```

Then test the model like this
```bash
./main \
  --model /path/to/ggml-model-q4_0.gguf \
  --threads 24 \
  --n_predict 2048 \
  --color \
  --interactive \
  --file prompts/chat-with-baichuan.txt \
  --reverse-prompt "用户:"
```

**Comment by jameswu2014** [comment]
> Besides the above fix, it would be better to provide a sample PROMPT file for Baichuan-13B-Chat.
> 
> Prompt file: `./prompts/chat-with-baichuan.txt`
> 
> ```
> 以下内容为人类用户与与一位智能助手的对话。
> 
> 用户:你好！
> 助手:
> ```
> 
> Then test the model like this
> 
> ```shell
> ./main \
>   --model /path/to/ggml-model-q4_0.gguf \
>   --threads 24 \
>   --n_predict 2048 \
>   --color \
>   --interactive \
>   --file prompts/chat-with-baichuan.txt \
>   --reverse-prompt "用户:"
> ```

good advices!

**Comment by MarvinLong** [comment]
python convert-baichuan-hf-to-gguf.py public/models/Baichuan-13B-Chat
 
 hello print:  BaichuanForCausalLM
gguf: found 3 model parts
num_parts:3

gguf: get model metadata
gguf: get tokenizer metadata
gguf: get sentencepiece tokenizer vocab, scores and token types
gguf: Setting special token type bos to 1
gguf: Setting special token type eos to 2
gguf: Setting special token type pad to 0
gguf: get tensor metadata
gguf: loading model part 'pytorch_model-00001-of-00003.bin'
Unpacking and permuting layer 0
Unpacking and permuting layer 1
Unpacking and permuting layer 2
Unpacking and permuting layer 3
Unpacking and permuting layer 4
Unpacking and permuting layer 5
Unpacking and permuting layer 6
Unpacking and permuting layer 7
Unpacking and permuting layer 8
Unpacking and permuting layer 9
Unpacking and permuting layer 10
Unpacking and permuting layer 11
Unpacking and permuting layer 12
Unpacking and permuting layer 13
Unpacking and permuting layer 14
--------------------------------------------------
It got stuck while I was converting the model.  Already 24 hours. And I tried again,  I got the same problem.
A800, Linux

**Comment by ggerganov** [comment]
@MarvinLong This looks something might be wrong with your data. Not sure, but it might be better to retry from scratch

**Review by ggerganov — APPROVED** [review]
Will be merging this soon.

Just a heads up - maintaining this model will primarily rely on contributions from the community. Adding some sort of CI in the future would help guarantee that the implementation is stable. But overall, if breaking changes occur, fixing Baichuan will be secondary priority

With time, we will try to refactor the code to reuse common building blocks when building the graphs of different models and this will probably help to keep everything stable together. I just want to get a few more architectures implemented before abstracting things

**Comment by jameswu2014** [comment]
> ```
>  python convert-baichuan-hf-to-gguf.py public/models/Baichuan-13B-Chat
>  
> 
> hello print:  BaichuanForCausalLM
> gguf: found 3 model parts
> num_parts:3
> 
> gguf: get model metadata
> gguf: get tokenizer metadata
> gguf: get sentencepiece tokenizer vocab, scores and token types
> gguf: Setting special token type bos to 1
> gguf: Setting special token type eos to 2
> gguf: Setting special token type pad to 0
> gguf: get tensor metadata
> gguf: loading model part 'pytorch_model-00001-of-00003.bin'
> Unpacking and permuting layer 0
> Unpacking and permuting layer 1
> Unpacking and permuting layer 2
> Unpacking and permuting layer 3
> Unpacking and permuting layer 4
> Unpacking and permuting layer 5
> Unpacking and permuting layer 6
> Unpacking and permuting layer 7
> Unpacking and permuting layer 8
> Unpacking and permuting layer 9
> Unpacking and permuting layer 10
> Unpacking and permuting layer 11
> Unpacking and permuting layer 12
> Unpacking and permuting layer 13
> Unpacking and permuting layer 14
> ```
> 
> It got stuck while I was converting the model. Already 24 hours. And I tried again, I got the same problem. A800, Linux

Is it solved? I will try it in my env.

**Comment by jameswu2014** [comment]
> 

Ok, Maybe we can help maintain baichuan model to be stable. For example, If you refactor the architecture, and that cause baichuan models do not work, we can help fix it. Actually, I think it will be great if  llama.cpp is  been refactored, there are many component that can be abstracted out. We are looking forward it.

**Comment by jameswu2014** [comment]
> Will be merging this soon.
> 
> Just a heads up - maintaining this model will primarily rely on contributions from the community. Adding some sort of CI in the future would help guarantee that the implementation is stable. But overall, if breaking changes occur, fixing Baichuan will be secondary priority
> 
> With time, we will try to refactor the code to reuse common building blocks when building the graphs of different models and this will probably help to keep everything stable together. I just want to get a few more architectures implemented before abstracting things

Ok, Maybe we can help maintain baichuan models to be stable. For example, If you refactor the architecture, and that cause baichuan models do not work, we can help fix it. Actually, I think it will be great if llama.cpp is been refactored, We are looking forward it.

**Comment by MarvinLong** [comment]
> from scratch



> > ```
> >  python convert-baichuan-hf-to-gguf.py public/models/Baichuan-13B-Chat
> >  
> > 
> > hello print:  BaichuanForCausalLM
> > gguf: found 3 model parts
> > num_parts:3
> > 
> > gguf: get model metadata
> > gguf: get tokenizer metadata
> > gguf: get sentencepiece tokenizer vocab, scores and token types
> > gguf: Setting special token type bos to 1
> > gguf: Setting special token type eos to 2
> > gguf: Setting special token type pad to 0
> > gguf: get tensor metadata
> > gguf: loading model part 'pytorch_model-00001-of-00003.bin'
> > Unpacking and permuting layer 0
> > Unpacking and permuting layer 1
> > Unpacking and permuting layer 2
> > Unpacking and permuting layer 3
> > Unpacking and permuting layer 4
> > Unpacking and permuting layer 5
> > Unpacking and permuting layer 6
> > Unpacking and permuting layer 7
> > Unpacking and permuting layer 8
> > Unpacking and permuting layer 9
> > Unpacking and permuting layer 10
> > Unpacking and permuting layer 11
> > Unpacking and permuting layer 12
> > Unpacking and permuting layer 13
> > Unpacking and permuting layer 14
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It got stuck while I was converting the model. Already 24 hours. And I tried again, I got the same problem. A800, Linux
> 
> Is it solved? I will try it in my env.

No, I tried buid from scratch 3 times, it always the same. I tried https://github.com/ouwei2013/baichuan13b.cpp/tree/master and it works on me. Can you give me a process to build from scratch? I can check whether I did something wrong.

**Comment by LMX-xin** [comment]
> 



> > from scratch
> 
> > > ```
> > >  python convert-baichuan-hf-to-gguf.py public/models/Baichuan-13B-Chat
> > >  
> > > 
> > > hello print:  BaichuanForCausalLM
> > > gguf: found 3 model parts
> > > num_parts:3
> > > 
> > > gguf: get model metadata
> > > gguf: get tokenizer metadata
> > > gguf: get sentencepiece tokenizer vocab, scores and token types
> > > gguf: Setting special token type bos to 1
> > > gguf: Setting special token type eos to 2
> > > gguf: Setting special token type pad to 0
> > > gguf: get tensor metadata
> > > gguf: loading model part 'pytorch_model-00001-of-00003.bin'
> > > Unpacking and permuting layer 0
> > > Unpacking and permuting layer 1
> > > Unpacking and permuting layer 2
> > > Unpacking and permuting layer 3
> > > Unpacking and permuting layer 4
> > > Unpacking and permuting layer 5
> > > Unpacking and permuting layer 6
> > > Unpacking and permuting layer 7
> > > Unpacking and permuting layer 8
> > > Unpacking and permuting layer 9
> > > Unpacking and permuting layer 10
> > > Unpacking and permuting layer 11
> > > Unpacking and permuting layer 12
> > > Unpacking and permuting layer 13
> > > Unpacking and permuting layer 14
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > It got stuck while I was converting the model. Already 24 hours. And I tried again, I got the same problem. A800, Linux
> > 
> > 
> > Is it solved? I will try it in my env.
> 
> No, I tried buid from scratch 3 times, it always the same. I tried https://github.com/ouwei2013/baichuan13b.cpp/tree/master and it works on me. Can you give me a process to build from scratch? I can check whether I did something wrong.
I tried it and it was able to run through baichuan-13B-chat.you should modify the convert-baichuan-hf-to-gguf.py ,and you can refer to the modifications above to modify the code.  https://github.com/ggerganov/llama.cpp/pull/3009#issuecomment-1706080220

File: convert-baichuan-hf-to-gguf.py [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+    model_part = torch.load(f"{dir_model}/{part_name}", map_location="cpu")
+
+    tmp=model_part
+    for i in itertools.count():
+        if f"model.layers.{i}.self_attn.W_pack.weight" in model_part:
+            print(f"Unpacking and permuting layer {i}")
+            tmp[f"model.layers.{i}.self_attn.q_proj.weight"]=reverse_hf_permute_part(model_part[f"model.layers.{i}.self_attn.W_pack.weight"],0,head_count,head_count)
+            tmp[f"model.layers.{i}.self_attn.k_proj.weight"]=reverse_hf_permute_part(model_part[f"model.layers.{i}.self_attn.W_pack.weight"],1,head_count,head_count_kv)
+            tmp[f"model.layers.{i}.self_attn.v_proj.weight"]=reverse_hf_part(model_part[f"model.layers.{i}.self_attn.W_pack.weight"],2)
+            del tmp[f"model.layers.{i}.self_attn.W_pack.weight"]
```
Comment (Cebtenzzre) [inline convert-baichuan-hf-to-gguf.py:241]: > No, I tried buid from scratch 3 times, it always the same. I tried [ouwei2013/baichuan13b.cpp@`master`](https://github.com/ouwei2013/baichuan13b.cpp/tree/master?rgh-link-date=2023-09-11T03%3A14%3A25Z) and it works on me. Can you give me a process to build from scratch? I can check whether I did something wrong.

I can also reproduce this. With either of these commands:
```
python convert-baichuan-hf-to-gguf.py ~/dirs/text-ai-models/dl/baichuan-inc_Baichuan-13B-Chat 0
```
```
python convert-baichuan-hf-to-gguf.py ~/dirs/text-ai-models/dl/baichuan-inc_Baichuan-13B-Chat 1
```
This infinite loop never ends, it keeps counting to insanely high values of `i`.

**Comment by LMX-xin** [comment]
> > from scratch
> 
> > > ```
> > >  python convert-baichuan-hf-to-gguf.py public/models/Baichuan-13B-Chat
> > >  
> > > 
> > > hello print:  BaichuanForCausalLM
> > > gguf: found 3 model parts
> > > num_parts:3
> > > 
> > > gguf: get model metadata
> > > gguf: get tokenizer metadata
> > > gguf: get sentencepiece tokenizer vocab, scores and token types
> > > gguf: Setting special token type bos to 1
> > > gguf: Setting special token type eos to 2
> > > gguf: Setting special token type pad to 0
> > > gguf: get tensor metadata
> > > gguf: loading model part 'pytorch_model-00001-of-00003.bin'
> > > Unpacking and permuting layer 0
> > > Unpacking and permuting layer 1
> > > Unpacking and permuting layer 2
> > > Unpacking and permuting layer 3
> > > Unpacking and permuting layer 4
> > > Unpacking and permuting layer 5
> > > Unpacking and permuting layer 6
> > > Unpacking and permuting layer 7
> > > Unpacking and permuting layer 8
> > > Unpacking and permuting layer 9
> > > Unpacking and permuting layer 10
> > > Unpacking and permuting layer 11
> > > Unpacking and permuting layer 12
> > > Unpacking and permuting layer 13
> > > Unpacking and permuting layer 14
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > It got stuck while I was converting the model. Already 24 hours. And I tried again, I got the same problem. A800, Linux
> > 
> > 
> > Is it solved? I will try it in my env.
> 
> No, I tried buid from scratch 3 times, it always the same. I tried https://github.com/ouwei2013/baichuan13b.cpp/tree/master and it works on me. Can you give me a process to build from scratch? I can check whether I did something wrong.

I tried it and it worked through baichuan-13B-chat. Please refer to the modification above to modify convert-baichuan-hf-to-gguf.py.   https://github.com/ggerganov/llama.cpp/pull/3009#issuecomment-1706080220

File: convert-baichuan-hf-to-gguf.py (hunk 1 above)
Comment (Cebtenzzre) [reply convert-baichuan-hf-to-gguf.py:241]: It looks like you misapplied the suggestion from https://github.com/ggerganov/llama.cpp/pull/3009#issuecomment-1706080220 - the loop should use `for i in range(block_count):`.

**Review by cebtenzzre — CHANGES_REQUESTED** [review]
The script currently does not work.

**Comment by jameswu2014** [comment]
> 

Could pull it again?

**Comment by jameswu2014** [comment]
> > from scratch
> 
> > > ```
> > >  python convert-baichuan-hf-to-gguf.py public/models/Baichuan-13B-Chat
> > >  
> > > 
> > > hello print:  BaichuanForCausalLM
> > > gguf: found 3 model parts
> > > num_parts:3
> > > 
> > > gguf: get model metadata
> > > gguf: get tokenizer metadata
> > > gguf: get sentencepiece tokenizer vocab, scores and token types
> > > gguf: Setting special token type bos to 1
> > > gguf: Setting special token type eos to 2
> > > gguf: Setting special token type pad to 0
> > > gguf: get tensor metadata
> > > gguf: loading model part 'pytorch_model-00001-of-00003.bin'
> > > Unpacking and permuting layer 0
> > > Unpacking and permuting layer 1
> > > Unpacking and permuting layer 2
> > > Unpacking and permuting layer 3
> > > Unpacking and permuting layer 4
> > > Unpacking and permuting layer 5
> > > Unpacking and permuting layer 6
> > > Unpacking and permuting layer 7
> > > Unpacking and permuting layer 8
> > > Unpacking and permuting layer 9
> > > Unpacking and permuting layer 10
> > > Unpacking and permuting layer 11
> > > Unpacking and permuting layer 12
> > > Unpacking and permuting layer 13
> > > Unpacking and permuting layer 14
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > It got stuck while I was converting the model. Already 24 hours. And I tried again, I got the same problem. A800, Linux
> > 
> > 
> > Is it solved? I will try it in my env.
> 
> No, I tried buid from scratch 3 times, it always the same. I tried https://github.com/ouwei2013/baichuan13b.cpp/tree/master and it works on me. Can you give me a process to build from scratch? I can check whether I did something wrong.

try again, Its my fault, Sorry.

File: convert-baichuan-hf-to-gguf.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
+#!/usr/bin/env python3
```
Comment (Cebtenzzre) [inline convert-baichuan-hf-to-gguf.py:1]: This file has a shebang, so it should be made executable (mode 0755).

File: convert-baichuan-hf-to-gguf.py [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+    return num_parts
+
+
+
+def parse_args() -> argparse.Namespace:
+    parser = argparse.ArgumentParser(description="Convert a HuggingFace LLaMA model to a GGML compatible file")
+    parser.add_argument("--vocab-only",  action="store_true",    help="extract only the vocab")
+    parser.add_argument("--outfile",     type=Path,              help="path to write to; default: based on input")
+    parser.add_argument("model",         type=Path,              help="directory containing model file, or model file itself (*.bin)")
+    parser.add_argument("ftype",     type=int, choices=[0, 1],   help="output format - use 0 for float32, 1 for float16", default = 1)
```
Comment (Cebtenzzre) [inline convert-baichuan-hf-to-gguf.py:64]: The default for 'ftype' does not work unless you also use `nargs='?'`.

**Comment by Cebtenzzre** [comment]
I can't use this script on a 13B model unless I set TMPDIR to disk. I have 24GB of RAM and 24GB of swap. Is this a fundamental limitation of these simpler convert scripts? I've never had such issues with 33B+ models and the standard convert.py.
```
Traceback (most recent call last):
  File "/home/cebtenzzre/src/forks/llama.cpp/convert-baichuan-hf-to-gguf.py", line 279, in <module>
    gguf_writer.add_tensor(new_name, data)
  File "/home/cebtenzzre/src/forks/llama.cpp/gguf-py/gguf/gguf.py", line 622, in add_tensor
    tensor.tofile(self.temp_file)
OSError: Not enough free space to write 140247040 bytes
```

File: convert-baichuan-hf-to-gguf.py (hunk 2 above)
Comment (jameswu2014) [reply convert-baichuan-hf-to-gguf.py:1]: got it

File: convert-baichuan-hf-to-gguf.py (hunk 3 above)
Comment (jameswu2014) [reply convert-baichuan-hf-to-gguf.py:63]: You are right, I have noticed this problem, We hope keep the consistency with other model converion script by now.Maybe we will fix it in the future.

**Comment by LMX-xin** [comment]
> I can't use this script on a 13B model unless I set TMPDIR to disk. I have 24GB of RAM and 24GB of swap. Is this a general limitation of these simpler convert scripts? I've never had such issues with 33B+ models and the standard convert.py.
> 
> ```
> Traceback (most recent call last):
>   File "/home/cebtenzzre/src/forks/llama.cpp/convert-baichuan-hf-to-gguf.py", line 279, in <module>
>     gguf_writer.add_tensor(new_name, data)
>   File "/home/cebtenzzre/src/forks/llama.cpp/gguf-py/gguf/gguf.py", line 622, in add_tensor
>     tensor.tofile(self.temp_file)
> OSError: Not enough free space to write 140247040 bytes
> ```

I tried and I ran it successfully.I used 19.4GB RAM to run the baichuan convert script.So it may be a general limitation.

**Comment by lx0126z** [comment]
How to inference by python after conversion?

**Comment by ggerganov** [comment]
> I can't use this script on a 13B model unless I set TMPDIR to disk.

The main `convert.py` script has some extra logic for lazy loading tensors, which allows to convert without loading the full model all at once. It's a bit over my head, but we probably will try to move this logic into `gguf.py` so it can be easily reused for other scripts. For now, the simple convert scripts will have this drawback in favor of simplicity

**Review by ggerganov — APPROVED** [review]
I see some questions are still pending. After resolving any issues - just merge it

**Review by cebtenzzre — APPROVED** [review]

**Pull request merged**